### PR TITLE
Refinements

### PR DESCRIPTION
**Pull request by bjfish**
Refinement specs/test results:
```
$ jt test spec/ruby/core/module/refine_spec.rb
1 file, 31 examples, 33 expectations, 0 failures, 0 errors, 0 tagged

$ jt test spec/ruby/core/module/using_spec.rb
1 file, 16 examples, 15 expectations, 0 failures, 0 errors, 1 tagged

$ jt test mri ruby/test_refinement.rb
63 tests, 324 assertions, 0 failures, 0 errors, 0 skips
```

TODO
- [x] add explanation of refinements implementation to `doc/contributor/refinements.md`
- [x] implement lexical scope added after calls to `using`

### CODE REVIEW

**Comment by chrisseaton** [comment]
Can you provide a couple of paragraphs explaining how it works? I can read the code but it's be nicer to be able to look at the algorithm more simply.

File: src/main/java/org/truffleruby/core/module/ModuleNodes.java [hunk 1]
```diff
@@ -1864,6 +1871,79 @@ public DynamicObject setMethodVisibility(DynamicObject module, Object name,
...
+
+            if (!RubyGuards.isRubyClass(classToRefine)) {
+                throw new RaiseException(coreExceptions().typeErrorWrongArgumentType(classToRefine, "Class", this));
+            }
+
+//            InternalMethod method = getContext().getCallStack().getCallingMethodIgnoringSend();
+//            LexicalScope lexicalScope = method == null ? null : method.getSharedMethodInfo().getLexicalScope();
+
+            ConcurrentMap<DynamicObject, DynamicObject> refinements = Layouts.MODULE.getFields(self).getRefinements();
+            ConcurrentMap<DynamicObject, DynamicObject> activatedRefinements = Layouts.MODULE.getFields(self).getRefinements();
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/module/ModuleNodes.java]: Should this be `getActivatedRefinements()`?

File: src/main/java/org/truffleruby/language/methods/AddMethodNode.java [hunk 2]
```diff
@@ -14,8 +14,11 @@
 import com.oracle.truffle.api.dsl.NodeChild;
 import com.oracle.truffle.api.dsl.NodeChildren;
 import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.VirtualFrame;
 import com.oracle.truffle.api.object.DynamicObject;
+import jnr.ffi.annotations.In;
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/methods/AddMethodNode.java]: Useless import?

File: src/main/java/org/truffleruby/core/module/ModuleFields.java [hunk 3]
```diff
@@ -79,9 +79,16 @@ public static void debugModuleChain(DynamicObject module) {
...
+    private boolean isRefinement = false;
+    private boolean isOverlaid = false;
+    private DynamicObject refinedClass;
+    private DynamicObject definedAt;
+
     private final ConcurrentMap<String, InternalMethod> methods = new ConcurrentHashMap<>();
     private final ConcurrentMap<String, RubyConstant> constants = new ConcurrentHashMap<>();
     private final ConcurrentMap<String, Object> classVariables = new ConcurrentHashMap<>();
+    private final ConcurrentMap<DynamicObject, DynamicObject> refinements = new ConcurrentHashMap<>();
+    private final ConcurrentMap<DynamicObject, DynamicObject> activatedRefinements = new ConcurrentHashMap<>();
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/module/ModuleFields.java]: What's activated refinements? Could you add a comment about it?
Also what is they key and the value, same for `refinements`.

File: src/main/ruby/core/main.rb [hunk 4]
```diff
@@ -42,6 +42,11 @@ def include(*mods)
     end
   end
 
+  def using(mod)
+    Truffle.invoke_primitive(:main_using, mod)
+    self
```
Comment (eregon) [inline src/main/ruby/core/main.rb]: Why a primitive and not a CoreMethod if the Ruby parts only calls it?

File: src/main/java/org/truffleruby/language/dispatch/CallDispatchHeadNode.java [hunk 5]
```diff
@@ -43,7 +43,8 @@ private CallDispatchHeadNode(boolean ignoreVisibility, boolean onlyCallPublic, M
     }
 
     public Object call(VirtualFrame frame, Object receiver, Object method, Object... arguments) {
-        return dispatch(frame, receiver, method, null, arguments);
+        // TODO BJF REVIEW Refinments probably apply to these calls as well
+        return dispatch(frame, receiver, method, null, null, arguments);
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/dispatch/CallDispatchHeadNode.java]: I think they don't, since these are calls in Java code, so typically in core methods, which are never in a refinement scope.

**Comment by bjfish** [comment]
@chrisseaton Here is my current explanation alongside a code example:
```
class A
  def one
    puts "one unrefined"
  end
end


module B
  refine A do               # Refining here will create an anonymous module R
  							# Module B now contains a refinements Map<RefinedClass, Refinement> to with an entry
  							# from class A to refinement module R.
  										
     def one                # This refine block is module eval'ed in the refinement module R.
       puts "one refined"   # Defining a method in the refinement module R will add the method to 
     end                    # to the refinement module and also update an existing matching method
                            # in refined class (A#one) to be a refined method entry
  end
end


using B  # This will iterate the entries defined in B's refinements Map<RefinedClass, Refinement>
         # and include the value Refinement module into the module chain for its key RefinedClass
         # 
         # It will also add the same refinement entries to the lexical scope 
         # so they are activated for the rest of the lexical scope.

A.new.one  # When a refined method entry is called, it will check the lexical scope to see if
           # the refinement module is currently active. 
           # If active, it will call the method in the refinement module. 
           # If not active, it will dispatch as normal without refinments.
```
An extended definition can be found in the spec:
https://bugs.ruby-lang.org/projects/ruby-trunk/wiki/RefinementsSpec

File: src/main/java/org/truffleruby/core/module/ModuleFields.java [hunk 6]
```diff
@@ -87,7 +87,11 @@ public static void debugModuleChain(DynamicObject module) {
     private final ConcurrentMap<String, InternalMethod> methods = new ConcurrentHashMap<>();
     private final ConcurrentMap<String, RubyConstant> constants = new ConcurrentHashMap<>();
     private final ConcurrentMap<String, Object> classVariables = new ConcurrentHashMap<>();
+
+    // Maps a refined class to refinement module
     private final ConcurrentMap<DynamicObject, DynamicObject> refinements = new ConcurrentHashMap<>();
+
+    // Maps a refined class to activated refinement module
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/module/ModuleFields.java]: What does `activated` mean? Is it caused by calling `using`?

File: src/main/java/org/truffleruby/core/module/ModuleFields.java [hunk 7]
```diff
@@ -87,7 +87,11 @@ public static void debugModuleChain(DynamicObject module) {
     private final ConcurrentMap<String, InternalMethod> methods = new ConcurrentHashMap<>();
     private final ConcurrentMap<String, RubyConstant> constants = new ConcurrentHashMap<>();
     private final ConcurrentMap<String, Object> classVariables = new ConcurrentHashMap<>();
+
+    // Maps a refined class to refinement module
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/module/ModuleFields.java]: Ah, so this is a list of refinements within a module wrapping `refine` calls, like `B` in your example.
Could you try to clarify?

Comment (bjfish) [reply src/main/java/org/truffleruby/core/module/ModuleFields.java]: This is where a modules refinements are actually stored so that when you call `using Mod` later they will be looked up here and added to the scope.  I've added an extended description of refinments here:
https://github.com/graalvm/truffleruby/pull/860#issuecomment-352157596

File: src/main/java/org/truffleruby/core/module/ModuleFields.java (hunk 6 above)
Comment (bjfish) [reply src/main/java/org/truffleruby/core/module/ModuleFields.java]: These are `activated` by calling refine. "Activated" means the module chain has been updated and the `activatedRefinements` are added to the lexical scope when yielding the refine body.

Normally, refinements are activated by calling `using` but the above refinements appear to be active using only refine for the purpose of adding them to the scope of the refine block.

File: src/main/ruby/core/main.rb (hunk 4 above)
Comment (bjfish) [reply src/main/ruby/core/main.rb]: There's some exception logic here that needs to be added:
https://github.com/ruby/ruby/blob/8dbc8caa0dfa366af6818b04e3ee4aebfe09478d/eval.c#L1467
I'm not sure yet how to do this top level check correctly.  It will probably make sense to move this to a core node.

File: src/main/java/org/truffleruby/core/module/ModuleNodes.java (hunk 1 above)
Comment (bjfish) [reply src/main/java/org/truffleruby/core/module/ModuleNodes.java]: Yes, fixed.

File: src/main/java/org/truffleruby/core/module/ModuleFields.java (hunk 3 above)
Comment (bjfish) [reply src/main/java/org/truffleruby/core/module/ModuleFields.java]: Commented added and more explanation: https://github.com/graalvm/truffleruby/pull/860#discussion_r157344825

File: src/main/ruby/core/main.rb (hunk 4 above)
Comment (bjfish) [reply src/main/ruby/core/main.rb]: I've updated this to use CoreMethod

**Comment by chrisseaton** [comment]
I thought it would use that lexical scope object we have - but it's actually applied to the module? Does that mean it's applied if you instance eval in the module?

**Comment by bjfish** [comment]
@chrisseaton We are using the LexicalScope object: https://github.com/graalvm/truffleruby/pull/860/files#diff-0de9a1ee6b4cc95802fb1124d1986549

The LexicalScope contains the refinement modules that have been activated inside that scope.
The Module contains the refinements modules defined by the refine call inside that module.

File: src/main/java/org/truffleruby/language/LexicalScope.java [hunk 8]
```diff
@@ -54,6 +59,19 @@ public static DynamicObject resolveTargetModuleForClassVariables(LexicalScope le
...
 
+    public Map<DynamicObject, DynamicObject> getRefinements() {
+        return refinements;
+    }
+
+    public Map<DynamicObject, DynamicObject> getRefinementsAllScopes() {
+        Map<DynamicObject, DynamicObject> results = new HashMap<>();
+        results.putAll(refinements);
+        if (this.getParent() != null) {
+            results.putAll(this.getParent().getRefinementsAllScopes());
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/LexicalScope.java]: This creates `N` maps and does O(n^2) `put()`.
Could you use iteration or a separate getRefinementsAllScopes(Map<> map) to avoid creating extra maps?

**Comment by bjfish** [comment]
I've added the update to move the refinements to  DeclarationContext. @eregon Please review the latest commit for this.

File: src/main/java/org/truffleruby/language/dispatch/UnresolvedDispatchNode.java [hunk 9]
```diff
@@ -96,9 +98,9 @@ public Object executeDispatch(
                             throw new UnsupportedOperationException();
                     }
                 } else if (RubyGuards.isRubyBasicObject(receiverObject)) {
-                    newDispatchNode = doDynamicObject(frame, first, receiverObject, methodName, argumentsObjects);
+                    newDispatchNode = doDynamicObject(frame, first, receiverObject, methodName, declarationContext, argumentsObjects);
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/dispatch/UnresolvedDispatchNode.java]: Do we need to pass the DeclarationContext explicitly here or could we just read it from the frame in `lookup()`?
If the caller frame is not a Ruby frame, refinements do not apply, just like visibility.

Comment (bjfish) [reply src/main/java/org/truffleruby/language/dispatch/UnresolvedDispatchNode.java]: @eregon I think this can be read from the frame now in some of these places instead of passing explicitly.

**Comment by bjfish** [comment]
@eregon I've added a document describing the implementation `doc/contributor/refinements.md`

File: src/main/java/org/truffleruby/language/dispatch/UnresolvedDispatchNode.java (hunk 9 above)
Comment (bjfish) [reply src/main/java/org/truffleruby/language/dispatch/UnresolvedDispatchNode.java]: @eregon I've updated this to read declaration context from the frame in the latest commit

File: doc/contributor/refinements.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+  refine C do  
+     def one
+       puts "one refined"
+     end
+  end
+end
+```
+The `refine` block is module eval'ed into a new anonymous module `R`. `R` is the refinement module and contains the method definitions from the refine block after they are eval'ed.  If the refined class `C` contains an existing method with the same name, the original method `one` in this example, the existing method will be marked as refined.
+
+Next, `refine` puts the new `R` module into module `M`'s refinements tables,  which is a map of refined classes to refinement modules. This specify entry will contain the key `C` (refined class) and the value `R` (anonymous refinement module).
```
Comment (eregon) [inline doc/contributor/refinements.md]: `specific`

File: doc/contributor/refinements.md [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+  
+  def two
+    C.new.one
+  end
+  
+end
+```
+The `using` method makes the refinements in module M active in the lexical scope that follows its call.  It does this by appending module `M`'s refinements table entries to the caller frame's declaration context.
+
+The `two` method in the `Test` module will save the frame's declaration context when defined.  When the `two` method is called, it will place it's declaration context into the frame so the refinements are seen in the method body.
```
Comment (eregon) [inline doc/contributor/refinements.md]: `its`
refinements are `applied`

File: src/main/java/org/truffleruby/core/module/ModuleFields.java [hunk 12]
```diff
@@ -79,10 +79,21 @@ public static void debugModuleChain(DynamicObject module) {
...
+    private boolean isOverlaid = false;
+    private DynamicObject refinedClass;
+    private DynamicObject definedAt;
+
     private final ConcurrentMap<String, InternalMethod> methods = new ConcurrentHashMap<>();
     private final ConcurrentMap<String, RubyConstant> constants = new ConcurrentHashMap<>();
     private final ConcurrentMap<String, Object> classVariables = new ConcurrentHashMap<>();
 
+    // Maps a refined class to refinement module
+    private final ConcurrentMap<DynamicObject, DynamicObject> refinements = new ConcurrentHashMap<>();
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/module/ModuleFields.java]: Is the key always the same as `rubyModuleObject` here?

Comment (eregon) [reply src/main/java/org/truffleruby/core/module/ModuleFields.java]: Nevermind, I misunderstood and clarified the comment.

**Comment by eregon** [comment]
Rebased on top of master. The diff is a bit simpler now.

**Comment by eregon** [comment]
I'm going through the failing specs one by one, but some of them are rather tough to implement.
The implementation is getting fairly clean.

Fun fact: we discovered MRI uses module chains as a poor man linked list in its refinements implementation.

**Comment by chrisseaton** [comment]
What's everyone's opinion on the state of this PR. There seems to be a trickle of fixes going in. Are they just fixing up more specs? Is the core of the work stable and ready to go in? I'd like to see Brandon merge his final PR before he leaves.

**Comment by bjfish** [comment]
I think it's ready to be merged.

**Comment by eregon** [comment]
All specs but one pass. I'll give it a quick shot later today, should be easy. Quite a few MRI tests fail though, some of them edge cases some maybe not so I'd like to review them.

**Comment by chrisseaton** [comment]
Ok, so those are just some specs which would be left to another PR? In that case I'll approve, as I'd like Brandon to be able to merge this PR since this is his last day.

**Comment by chrisseaton** [comment]
Funny how it ended up being less than 800 lines. Did we merge in some other work previously to prepare for refinements, or is this all of it?

**Comment by chrisseaton** [comment]
@eregon wants to try to fix the one last spec today.

**Comment by eregon** [comment]
> Funny how it ended up being less than 800 lines. Did we merge in some other work previously to prepare for refinements, or is this all of it?

We merged #926 and #932 which are general fixes to `DeclarationContext`, which is the base to keep track of currently active refinements.
But indeed, refinements are not so complicated to implement in the end.

File: src/main/java/org/truffleruby/parser/BodyTranslator.java [hunk 13]
```diff
@@ -721,7 +721,9 @@ private RubyNode translateCallNode(CallParseNode node, boolean ignoreVisibility,
 
         final SourceIndexLength enclosingSourceSection = enclosing(sourceSection, children.toArray(new RubyNode[children.size()]));
 
-        RubyCallNodeParameters callParameters = new RubyCallNodeParameters(receiver, methodName, argumentsAndBlock.getBlock(), argumentsAndBlock.getArguments(), argumentsAndBlock.isSplatted(), privately || ignoreVisibility, isVCall, node.isLazy(), isAttrAssign);
+        RubyCallNodeParameters callParameters = new RubyCallNodeParameters(receiver, methodName, argumentsAndBlock.getBlock(), argumentsAndBlock.getArguments(), argumentsAndBlock.isSplatted(), privately || ignoreVisibility, isVCall, node.isLazy(), isAttrAssign,
+                null); //getLexicalScopeNode("call method", sourceSection)
```
Comment (bjfish) [inline src/main/java/org/truffleruby/parser/BodyTranslator.java]: Remove commented out code

File: src/main/java/org/truffleruby/language/methods/UsingNode.java [hunk 14]
```diff
@@ -0,0 +1,78 @@
+package org.truffleruby.language.methods;
```
Comment (bjfish) [inline src/main/java/org/truffleruby/language/methods/UsingNode.java]: Needs copyright

File: src/main/java/org/truffleruby/language/methods/LookupMethodNode.java [hunk 15]
```diff
@@ -158,18 +160,18 @@ protected DynamicObject metaClass(Object object) {
     }
 
     protected MethodLookupResult doCachedLookup(VirtualFrame frame, Object self, String name) {
-        return lookupMethodCachedWithVisibility(getContext(), frame, self, name, ignoreVisibility, onlyLookupPublic);
+        return lookupMethodCachedWithVisibility(getContext(), frame, self, name, ignoreVisibility, onlyLookupPublic); // TODO BJF Fix cached lookup
```
Comment (bjfish) [inline src/main/java/org/truffleruby/language/methods/LookupMethodNode.java]: @eregon I think you worked on the cached lookup and this `TODO` could be removed?

File: src/main/java/org/truffleruby/language/methods/UsingNode.java (hunk 14 above)
Comment (eregon) [reply src/main/java/org/truffleruby/language/methods/UsingNode.java]: OK, I'll add it.

File: src/main/java/org/truffleruby/language/methods/LookupMethodNode.java (hunk 15 above)
Comment (eregon) [reply src/main/java/org/truffleruby/language/methods/LookupMethodNode.java]: Indeed, looks done to me, removed.

File: src/main/java/org/truffleruby/language/methods/AddMethodNode.java [hunk 16]
```diff
@@ -44,18 +46,38 @@ public void addMethod(DynamicObject module, InternalMethod method, Visibility vi
...
+    }
 
+    @TruffleBoundary
+    private void addRefinedMethodEntry(DynamicObject module, InternalMethod method, Visibility visibility) {
+        final MethodLookupResult result = ModuleOperations.lookupMethodCached(module, method.getName(), null);
+        final InternalMethod originalMethod = result.getMethod();
+        if (originalMethod == null) {
+            doAddMethod(module, method.withRefined(true).withOriginalMethod(null), visibility);
+        } else if (originalMethod.isRefined()) {
+            // Already marked as refined
```
Comment (bjfish) [inline src/main/java/org/truffleruby/language/methods/AddMethodNode.java]: Maybe an edge case: I could imagine a scenario where the same method is refined twice so the original method maybe incorrect then.  Could handle this in a later PR though.

**Comment by eregon** [comment]
@bjfish I think this PR is now ready to be merged :smiley:
